### ISSUE TITLE
[CTS 5] Fallback fonts (font glyph substitution)

### ISSUE THREAD

**Issue by PhilterPaper** [PhilterPaper/Perl-PDF-Builder#56]
Subject: CTS 5 - Fallback fonts (font glyph substitution)
March 17, 2017, 04:40:39 AM by sciurius

I'm not sure if this is the right thread, but anyway...

In this modern Unicode era we run into the problem that there are many more symbols to be shown than are present in a particular font. For example, when dealing with music I want to show a sharp ♯, flat ♭, delta Δ, and so on. I notice that many applications, like the web browser I'm typing this in, use font glyph substitution using fallback fonts if necessary. The glyphs are borrowed from a different font. Google has developed free fonts (the Noto-fonts) to complement its standard font (Roboto) with virtually all glyphs defined in Unicode.

I think  supporting fallback fonts is a must for a future-proof PDF::Builder.
#
March 17, 2017, 09:24:13 AM by Phil

Indeed, this would be the right place to request this feature, and it's a good one. Browsers have done this for quite a while, first looking for a glyph in the first requested font-family, then the second, etc. until the default fallback. Actually, come to think of it, I'm not sure what happens if you explicitly list one or more font-families in CSS and the glyph is not found in any of them — does a browser try its own list of fonts in a desperate attempt to find _something_ instead of a tofu? For PDF generation, that might be something configurable.

What is the industry standard for mapping a requested (but missing) glyph to something else, especially if they don't happen to have the same CId or Unicode code point or a standard name? The original encoding would say what the character is, but in a fallback font it could be at a different code point or under a different name or ID.
#
March 17, 2017, 03:00:43 PM by sciurius

Industry standard:

HarfBuzz is an OpenType text shaping engine.

The current HarfBuzz codebase [...] is stable and under active maintenance. This is what is used in latest versions of Firefox, GNOME, ChromeOS, Chrome, LibreOffice, XeTeX, Android, and KDE, among other places.

**Comment by PhilterPaper** [comment]
See also http://www.catskilltech.com/forum/pdf-builder-general-discussions/font-families/. This is a somewhat related discussion for PDF::Builder knowing what font variants are available for each supported font, and being able to specify font variants (bold, italic, small caps, etc.) in a simple and consistent way. It might even involve going to another font, or building a "synthetic font", to get unsupported variants. Someone would have to specify a list of fallback fonts to look at, before either crudely synthesizing a variant (e.g., font size change for small caps, or overprinting (with offset) for bold) or just giving up.

**Comment by PhilterPaper** [comment]
One thing to keep in mind is that a requested glyph is going to be known only by its code point and encoding (xFF in Latin-1 is different than xFF in CP1253, and isn't valid in UTF-8). At that point (knowing the output encoding of a text string), each character could be checked against a list of supported glyphs for a font, and a request to generate an alternate font could be made. Unfortunately, this sounds like it could be quite costly in time and/or space, but that may be the price to pay, especially if not all operating systems offer utilities to do this.

**Comment by terefang** [comment]
have you looked into unifont.pm ? this is what this module is for.

**Comment by PhilterPaper** [comment]
I don't think this is what `unifont()` is intended for. As far as I can tell, `unifont` specifies that a certain opened font is to be used for a certain range of Unicode points. The request made by Johan is for font **fallback**, if a requested Unicode point does _not_ have a glyph available in the current font. A browser does this, looking down a list of desired typefaces (font families) until a glyph is found. 

I suppose that if you know _in advance_ that a desired glyph will be unavailable in your chosen font, that you could select (via `unifont`) an alternate font to use for that Unicode point, but that's not quite what was requested (and involves a lot more manual work).

@terefang thanks for joining in the discussion, and I hope to see more input from you!

**Comment by terefang** [comment]
@PhilterPaper 

yes, you are right that (with plain unifont.pm) you have to know in advance the codepoints to fallback to.

nonetheless are the introspection capabilities of the font objects good enough for this, so the code to be written wont get too complicated.

the real question would be which font to fallback to.

**Comment by PhilterPaper** [comment]
The original request (from Johan) is for HTML/CSS style to go through a list of fonts (let's assume italic and bold are dealt with properly), in the indicated order, until a font is found that has a glyph for the desired Unicode point. In other words, give a list of most desirable to least desirable typefaces/fonts and the Unicode text to set, and hopefully most of the time you'll get glyphs from the most desired typeface, with occasional fallbacks to the less desired typefaces.

Anyway, it shouldn't be _terribly_ hard to do such a thing. The code just needs to detect that there is no glyph (CId) for the Unicode character, and (rather than outputting a blank or an empty box) go down the list of alternate fonts, opening up the appropriate ones and checking if the desired glyph exists. This might be a good extension to Johan's `Text::Layout`, which already concerns itself with keeping track of the desired typeface and whether it's bold and/or italic (among other things). IIRC you have to pre-open all the fonts you'll use, so worst case you'd have to open some more for alternates.

While we're at it, the `unifont()` method of specifying different lists of specific opened fonts for ranges of Unicode characters (single-byte encodings too?) might be blended in.

**Comment by terefang** [comment]
you can basically query a truetype font if a particular unicode-point is available (via cmap).

since there are so many bad or broken fonts out there, i never bothered to implemented it in `unifont()`, 
but rather have the user specify what she needed.

as for assembling missing glyphs from a combination, this only works for a very systematically contructed font but tends to break down very rapidly otherwise. (i used the equivalent in fontforge and 90% of the time needed to do tweaking)

**Comment by PhilterPaper** [comment]
(Let me cc @sciurius on this in case he has anything to add to the discussion. He mentioned that HarfBuzz::Shaper might have some utility for finding a font with a given Unicode point. I'm not sure he's put that into Text::Layout.)

For a font which has already been loaded, a method such as `glyphByUni($UnicodePoint)` might work (returns `.notdef` if this code point is missing). The question is whether you want to go ahead and load possibly a large number of other fonts, to find the first with the desired code point present. It's not a horrendous burden on the system, but not a great thing to have happen. We would have to make sure that a rejected font (not used) doesn't get loaded into the PDF.

Can we discover code point availability _before_ loading in a given font? Something would be needed to read a given font file (presumably TTF/OTF) and see what's in it, without formally loading it into PDF::Builder (though that _is_ possible). We already prereq Font::TTF, I wonder if it has any capabilities here?

**Comment by sciurius** [comment]
As far as I know, Font::TTF already delays loading the font tables so it should be possible to inspect the relevant tables without 'loading everything'.

**Comment by PhilterPaper** [comment]
Another thread on the subject:  sciurius/perl-Text-Layout#6

**Comment by PhilterPaper** [comment]
Just thinking about the practical aspects of doing fallbacks (looking in another font if the requested glyph is missing from this one). First, the lookup should be _cached_ so that if a character is requested a second time, its fallback (or primary) can be rapidly selected. Second, can the ASCII character set be _assumed_ to be present in the current font? I suppose that there _could_ be specialty fonts (especially decorative or symbol ones) with only a limited subset of ASCII in them. Third, the glyphs would have to be embedded in the PDF file (probably a good idea anyway), as we can't depend on an external font file to have the same glyphs, or to be in the same place! Where does this leave core and PS/T1 fonts? How about non-UTF8 encodings? Can we depend on non-TTF fonts (not embedded) always having the same glyph set? 

Font Manager can be used to easily switch between fonts when looking for the glyph for a Unicode point. I presume that at some point in the search, the font would actually have to be loaded, in order to inquire whether there's a glyph for that point. If a font in the fallback list ends up not being used at all, there should be a way to automatically discard it (assuming it would otherwise be incorporated into the PDF file).

**Comment by terefang** [comment]
let me add my thoughts ... in no particular order.

* first and foremost – not embedding a font is bad as it breaks PDF/A and PDF/X assumptions and users usually dont share their fonts with their PDFs.
* give the user/programmer the option to always embed the fallback fonts (probably subset).
* if the fallback font is less than 100kb, embed it with flate compression as nobody will care to be bothered.
* if only a few glyphs are used for fallback (<256), you can create a synthetic Type3 font out of the glyphset – this will also allow you to combine glyphs from multiple fallback fonts.
* if the fallback font is also referenced as a normal font the need to embed it is clear.
* if a fallback font is reference multiple times, either embed it (subset) or got the Type3 route.
* you basically cannot assume a basic ASCII glyphset to be present, except in really old broken 90s ttfs.

most usecases that i have encountered will : 
* either combine multiple encoded fonts to form a kind of Unicode font
* combine an ASCII/Greek/Cyrillic encoded font with an Unicode fallback font
* combine a writing font with an icon font (like dings or awesome)

hope that helped

**Comment by terefang** [comment]
T1/T3 fonts are encoded in a single-byte encoding that might or might not include ASCII in 0x20-0x7F (ie depending on whats included in the AFM/PFM file).

for TTF/OTF that depends on their embedded CMaps if ASCII glyphs are available or not. if a Symbol CMap is present like it is for WingDings and many newer iconic fonts, chances are pretty high that ASCII glyphs are not present.

in the 90s, people would use various bad font tools to reencode such fonts (including me) to bring their glyphs into a single-byte range via (duplicate) compatiblity entries in the CMap. (eg. mapping 0xE*?? or 0xF*?? to 0x00??)

it basically becomes a 80:20 gamble ... the standard code path with default options should be sufficient for 80% of the usecases. the remaining 20% will need to set proper options for the processing and maybe need to deep-dive into the fonts they use and determine if they are new/good enough or so crappy that they need to be replaced.

i have also seen the `Nerd Fonts` Project which simply patches entire font collections to include missing glyphs. (ie in their case icons)

**Comment by terefang** [comment]
@PhilterPaper i just read the following test in the PDF1.7 spec:

> The results of the CMap mapping algorithm are a font number and a character selector. The font number shall
be used as an index into the Type 0 font’s DescendantFonts array to select a CIDFont. In PDF, the font
number shall be 0 and the character selector shall be a CID; this is the only case described here. The CID shall
then be used to select a glyph in the CIDFont. *(PDF 32000-1:2008 – PDF 1.7 – page 280)*

do i interpret that correctly that previously it was possible to construct a Type0 font with multiple DescendantFonts and one could concstruct a CMap to select the CIDs from it ?

also reading section 6 of 5014.CIDFont_Spec would suggest so.

**Comment by terefang** [comment]
just verified this in Poppler source code that it only uses the first entry in the array

**Comment by PhilterPaper** [comment]
> do i interpret that correctly that previously it was possible to construct...

You're asking _me_? I know nothing about such things :-) -- I'm counting on people like _you_ to be the experts!

**Comment by terefang** [comment]
> > do i interpret that correctly that previously it was possible to construct...
> 
> You're asking _me_? I know nothing about such things :-) -- I'm counting on people like _you_ to be the experts!

i asked around in my local commune of STEM people and we have come to the conclusion that:

* it should be possible in Postscript
* it might have been planned for inclusion in PDF
* the language of the ISO-32000 spec suggests a deprecation of the feature in PDF
* looking at several pdf sources seem to confirm the third point

this is a sad day for me ... that would have been a very useful feature

**Comment by PhilterPaper** [comment]
As (at least) an interim step, if determine that a glyph (or entire font?) is missing, replace with a "tofu" per character, and a single warning to STDERR. See discussion in #214 about implementing 'tofu'.  To actually replace with existing glyphs remains to be seen... it will probably have to be defined _manually_ on a glyph-by-glyph basis, possibly including synthesizing glyphs on the fly (e.g., triple circular integral, or compositing a LaTeX logo, although _that_ has no Unicode codepoint), as well as substituting from another font. Maybe an internal table: "if missing this glyph for these (or **all**) font(s), use XXX from this font". E.g., for Greek characters (Unicode range) go to the Symbol font and translate to this codepoint range. Missing variants (see next paragraph) could be handled with `synfont()`. This table would have to be manually maintained by the user.

Eventually FontManager will have a way to define handling missing _font_ variants (e.g., requested Bold Symbol), including calling `synfont()` on the fly and mapping to existing variants, which will take care of entire missing fonts (to a large degree). Handling completely missing type _faces_, I don't know about, maybe something similar to missing glyphs?

**Comment by PhilterPaper** [comment]
#214 discusses outputting a "tofu" box if a glyph can't be found (per this ticket, after searching the entire list of fallback fonts). Also see if HarfBuzz does anything about looking for fallbacks, as it already can do a lot of glyph substitutions.

**Comment by terefang** [comment]
for an approach see here:

https://www.mrumpler.at/the-trouble-with-text-rendering-in-skiasharp-and-harfbuzz/#font-fallback

**Comment by PhilterPaper** [comment]
Hi Alfred! 

PDF::Builder already offers HarfBuzz shaping (via HarfBuzz::Shaper package), although it's not yet fully integrated into text processing (e.g, not yet into `column()`). It looks like the "font fallback" code in your link tries to find a font that covers _all_ the desired glyphs, rather than sticking to one font as much as possible, and exploring other fonts only for individual missing glyphs. I suppose that either approach could be desired by a user. How big a chunk of text do you check that the desired font has all the glyphs? The entire 1000 page document? Maybe that should be left to the user to manually handle. Just remind them at the end that tofu(s) were output (give page list), and they might want to check the font used. In a pinch, slanted and overstruck (poor man's bold) might be used if the Roman/regular version has the desired glyph but italic or bold doesn't.

I think it would be reasonable for PDF::Builder to support not just a single font family (face) at a time, but support a _list_ of font families to try as alternatives, much as HTML/CSS can do (my understanding is that _available_ alternate fonts are explored on an as-needed per-character basis, rather than the whole phrase covered by the font-family, but I haven't explored it deeply).

BTW, PDF::Builder will **never** support Emojis. I don't care how popular they have become for poorly educated, mush-brained Gen Z -- we have a perfectly serviceable alphabet developed over thousands of years, and that's good enough for me! :-)

**Comment by sciurius** [comment]
> BTW, PDF::Builder will never support Emojis. I don't care how popular they have become for poorly educated, mush-brained Gen Z

Emojis have significantly simplified my spam filters. I have yet to see a mail with an emoji in its subject that was not spam or otherwise irrelevant ;)

**Comment by PhilterPaper** [comment]
> my understanding is that available alternate fonts are explored on an as-needed per-character basis, rather than the whole phrase covered by the font-family

Per https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/font-family, that appears to be the way it works. Start with the first available font, and for any glyph not found, try the other fonts, returning to the first font each time for the next glyph.

> The `font-family` property specifies a list of fonts, from highest priority to lowest. Font selection _does not_ stop at the first font in the list that is on the user's system. Rather, font selection is done _one character at a time_, so that if an available font does not have a glyph for a needed character, the latter fonts are tried. When a font is only available in some [styles](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/font-style), [variants](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/font-variant), or [sizes](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/font-size), those properties may also influence which font family is chosen.

I suppose there _could_ be a mode to try to pick a single font for an entire phrase (however much you define that to be), but that could be slower than going along with the first chosen font until you hit a missing glyph, _then_ exploring other fonts in the list, and as a last resort, outputting a tofu. Perhaps it would be desirable to give a warning about "involuntary font change due to missing glyphs" page list (just as with the use of tofus), and let the user decide what they want to do about it.
